---
- hosts: all
  become: true
  vars:
    mount_location: '/opt/test1'
    testfile: "{{ mount_location }}/quux"
    volume_size: '5g'

  tasks:
    - include_role:
        name: linux-system-roles.storage

    - include_tasks: get_unused_disk.yml
      vars:
        min_size: "{{ volume_size }}"
        max_return: 1

    ##
    ## Disk
    ##
    - name: Test for correct handling of new encrypted volume w/ no key
      block:
        - name: Create an encrypted disk volume w/ default fs
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_volumes:
              - name: foo
                type: disk
                disks: "{{ unused_disks }}"
                mount_point: "{{ mount_location }}"
                encryption: true

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the keyless luks test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
                   not blivet_output.changed"
            msg: "Unexpected behavior w/ encrypted pool w/ no key"

    # encrypted disk volume
    - name: Create an encrypted disk volume w/ default fs
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_volumes:
          - name: foo
            type: disk
            disks: "{{ unused_disks }}"
            mount_point: "{{ mount_location }}"
            encryption: true
            encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Remove the encryption layer
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_volumes:
              - name: foo
                type: disk
                disks: "{{ unused_disks }}"
                mount_point: "{{ mount_location }}"
                encryption: false
                encryption_password: 'yabbadabbadoo'
        - name: unreachable task
          fail:
            msg: UNREACH
      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to encryption removal')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing filesystem in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Remove the encryption layer
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_volumes:
          - name: foo
            type: disk
            disks: "{{ unused_disks }}"
            mount_point: "{{ mount_location }}"
            encryption: false
            encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Add encryption to the volume
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_volumes:
              - name: foo
                type: disk
                disks: "{{ unused_disks }}"
                mount_point: "{{ mount_location }}"
                encryption: true
                encryption_password: 'yabbadabbadoo'
        - name: unreachable task
          fail:
            msg: UNREACH
      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to adding encryption')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing filesystem in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Add encryption to the volume
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_volumes:
          - name: foo
            type: disk
            disks: "{{ unused_disks }}"
            mount_point: "{{ mount_location }}"
            encryption: true
            encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    ##
    ## Partition
    ##

    - name: Test for correct handling of new encrypted volume w/ no key
      block:
        - name: Create an encrypted partition volume w/ default fs
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_safe_mode: false
            storage_pools:
              - name: foo
                type: partition
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    type: partition
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: true

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the keyless luks test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
                   not blivet_output.changed"
            msg: "Unexpected behavior w/ encrypted pool w/ no key"

    - name: Create an encrypted partition volume w/ default fs
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: partition
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                type: partition
                mount_point: "{{ mount_location }}"
                size: 4g
                encryption: true
                encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Remove the encryption layer
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_pools:
              - name: foo
                type: partition
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    type: partition
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: false
                    encryption_password: 'yabbadabbadoo'
        - name: unreachable task
          fail:
            msg: UNREACH
      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to encryption removal')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing filesystem in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Remove the encryption layer
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: partition
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                type: partition
                mount_point: "{{ mount_location }}"
                size: 4g
                encryption: false
                encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Add encryption to the volume
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_pools:
              - name: foo
                type: partition
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    type: partition
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: true
                    encryption_password: 'yabbadabbadoo'

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to adding encryption')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing volume in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Test key file handling
      block:
        - name: Create a key file
          tempfile:
            prefix: storage_test
            suffix: lukskey
          register: storage_test_key_file
          changed_when: false

        - name: Write the key into the key file
          copy:
            dest: "{{ storage_test_key_file.path }}"
            content: veryveryveryveryverysecureindeed
            mode: "0600"
          changed_when: false

        - name: Add encryption to the volume
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_safe_mode: false
            storage_pools:
              - name: foo
                type: partition
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    type: partition
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: true
                    encryption_key: "{{ storage_test_key_file.path }}"

        - include_tasks: verify-role-results.yml
      always:
        - name: Remove the key file
          file:
            path: "{{ storage_test_key_file.path }}"
            state: absent
          changed_when: false

    ##
    ## LVM
    ##

    - name: Test for correct handling of new encrypted volume w/ no key
      block:
        - name: Create an encrypted lvm volume w/ default fs
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_safe_mode: false
            storage_pools:
              - name: foo
                type: lvm
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: true

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the keyless luks test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg|regex_search('encrypted volume.*missing key') and
                   not blivet_output.changed"
            msg: "Unexpected behavior w/ encrypted pool w/ no key"

    - name: Create an encrypted lvm volume w/ default fs
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: lvm
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                mount_point: "{{ mount_location }}"
                size: 4g
                encryption: true
                encryption_password: 'yabbadabbadoo'
                encryption_luks_version: luks1
                encryption_key_size: 512
                encryption_cipher: 'serpent-xts-plain64'

    - include_tasks: verify-role-results.yml

    - name: Verify preservation of encryption settings on existing LVM volume
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: lvm
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                mount_point: "{{ mount_location }}"
                size: 4g

    - name: Assert preservation of encryption settings on existing LVM volume
      assert:
        that: not blivet_output.changed and
              blivet_output.pools[0].volumes[0].encryption and
              blivet_output.pools[0].volumes[0].encryption_luks_version == 'luks1'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Remove the encryption layer
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_pools:
              - name: foo
                type: lvm
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: false
                    encryption_password: 'yabbadabbadoo'

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to encryption removal')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing volume in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Remove the encryption layer
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: lvm
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                mount_point: "{{ mount_location }}"
                size: 4g
                encryption: false
                encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - import_tasks: create-test-file.yml

    - name: Test for correct handling of safe_mode
      block:
        - name: Add encryption to the volume
          include_role:
            name: linux-system-roles.storage
          vars:
            storage_pools:
              - name: foo
                type: lvm
                disks: "{{ unused_disks }}"
                volumes:
                  - name: test1
                    mount_point: "{{ mount_location }}"
                    size: 4g
                    encryption: true
                    encryption_password: 'yabbadabbadoo'

        - name: unreachable task
          fail:
            msg: UNREACH

      rescue:
        - name: Check that we failed in the role
          assert:
            that:
              - ansible_failed_result.msg != 'UNREACH'
            msg: "Role has not failed when it should have"

        - name: Verify the output of the safe_mode test
          assert:
            that: "blivet_output.failed and
                   blivet_output.msg
                   |regex_search('cannot remove existing
                   formatting.*in safe mode due to adding encryption')
                   and not blivet_output.changed"
            msg: "Unexpected behavior w/ existing volume in safe mode"

    - import_tasks: verify-data-preservation.yml

    - name: Add encryption to the volume
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_safe_mode: false
        storage_pools:
          - name: foo
            type: lvm
            disks: "{{ unused_disks }}"
            volumes:
              - name: test1
                mount_point: "{{ mount_location }}"
                size: 4g
                encryption: true
                encryption_password: 'yabbadabbadoo'

    - include_tasks: verify-role-results.yml

    - name: Clean up
      include_role:
        name: linux-system-roles.storage
      vars:
        storage_volumes:
          - name: foo
            type: disk
            disks: "{{ unused_disks }}"
            state: absent

    - include_tasks: verify-role-results.yml
